File: Lab3/Lab3/Character.cpp
#define _CRT_SECURE_NO_WARNINGS
#include "Character.h"
#include <iostream>
#include <ctime>

int Character::counter = 0;

Character::Character(const std::string& name, const std::string& race, int level, Equipment* left, Equipment* right)
{
    this->name = name;
    this->race = race;
    this->level = level;
    left_hand = left;
    right_hand = right;
    id = counter++;
    update_data();
}

Character::Character(const std::string& name)
{
    this->name = name;
    race = races[std::rand() % races.size()];
    level = std::rand() % 10 + 1;
    left_hand = new Equipment(std::rand() % 10 + 1);
    right_hand = new Equipment(std::rand() % 10 + 1);
    id = counter++;
    update_data();
}

Character::Character()
{
    name = names[std::rand() % names.size()];
    race = races[std::rand() % races.size()];
    level = std::rand() % 10 + 1;
    left_hand = new Equipment(std::rand() % 10 + 1);
    right_hand = new Equipment(std::rand() % 10 + 1);
    id = counter++;
    update_data();
}

Character::Character(const Character& character) : id(character.id), name(character.name), race(character.race), level(character.level), left_hand(new Equipment(*character.left_hand)), right_hand(character.right_hand)
{
    update_data();
}

int Character::get_counter()
{
    return counter;
}

std::string Character::read()
{
    std::cout << "Input character name: ";
    std::string s;
    std::cin >> s;
    return s;
}

void Character::display(Character* character)
{
    character->display();
}


void Character::display()
{
    std::cout << "ID: "<< id << "\nName: " << name << "\nRace: " << race << "\nLevel: " << level << "\nExperience: " << experience << " \/ 100" << "\nHealth: " << health << "\nMana: " << mana << "\nDamage: " << get_damage() << " (" << damage << "+" << get_equipment_damage() << ")"   << "\n";
    if (left_hand != nullptr && left_hand->get_name() != "")
    {
        std::cout << left_hand->get_name() << ": " << left_hand->get_damage() << std::endl;
    }
    if (right_hand != nullptr && right_hand->get_name() != "")
    {
        std::cout << right_hand->get_name() << ": " << right_hand->get_damage() << std::endl;
    }
    std::cout << std::endl;
}


void Character::level_up()
{
    level++;
    update_data();
}

void Character::add(int exp)
{
    experience += exp;
    while (experience >= 100)
    {
        experience -= 100;
        level_up();
    }
}


void Character::take_damage(int dmg)
{
    health -= dmg;
    if (health <= 0)
    {
        health = 40 * level;
        experience = 0;
        dead = true;
        std::cout << "You're dead" << std::endl;
    }
}


void Character::update_data()
{
    health = 40 * level;
    mana = 20 * level;
    damage = 5 * level;
}

std::string Character::get_race()
{
    return race;
}

int Character::get_equipment_damage()
{
    if (left_hand != nullptr && right_hand != nullptr)
        return ((*left_hand) + (*right_hand)).get_damage();
    return 0;
}

int Character::get_damage()
{
    return damage + get_equipment_damage();
}

int Character::get_experience()
{
    return experience;
}

int Character::get_health()
{
    return health;
}

int Character::get_level()
{
    return level;
}

int Character::get_mana()
{
    return mana;
}

Equipment* Character::get_left_hand()
{
    return left_hand;
}
Equipment& Character::get_right_hand() 
{
    return *right_hand;
}

File: Lab6/Character.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Lab6
{
    class Character
    {
        
        private string name;
        private string race;
        private int level = 0;
        private int health = 0;
        private int mana = 0;
        private int experience = 0;
        private float damage = 0;
        private bool dead = false;
        private Random random = new Random();
        public static List<string> NAMES = new List<string> { "Lo Whitbottom", "Meginhard Swiftfoot", "Godomar Gawkroger", "Gerontius Chubb - Baggins", "Walcaud Leafwalker", "Griffon Harfoot", "Balbo Sackville", "Butilin Lothran", "Willehad Silverstring", "Berengar Undertree" };
        public static List<string> RACES = new List<string> { "Human", "Orc", "Elf", "Hobbit", "Charr", "Norn" };
        private Equipment leftHand = null;
        private Equipment rightHand = null;

        public void Init(string name)
        {
            this.name = name;
            race = "Human";
            level = 1;
            health = 40;
            mana = 20;
            damage = 5;
            experience = 0;
        }

        public void Init()
        {
            name = NAMES[random.Next(NAMES.Count)];
            race = RACES[random.Next(RACES.Count)];
            level = random.Next(10) + 1;
            leftHand = new Equipment(random.Next(25) + 1);
            rightHand = new Equipment(random.Next(25) + 1);
            UpdateData();
        }

        public string Read()
        {
            Console.WriteLine("Input character name: ");
            return Console.ReadLine();
        }

        public void Display()
        {
            Console.WriteLine(string.Format("Name: {0}\nRace: {1}\nLevel: {2}\nExperience: {3} / 100\nHealth: {4}\nMana: {5}\nDamage: {6} ({7} + {8})\n\n", name, race, level, experience, health, mana, GetDamage(), damage, GetEquipmentDamage()));
        }

        public void LevelUp()
        {
            level++;
            UpdateData();
        }

        public void Add(int exp)
        {
            experience += exp;
            while (experience >= 100)
            {
                experience -= 100;
                LevelUp();
            }
        }

        public void TakeDamage(int dmg)
        {
            health -= dmg;
            if (health <= 0)
            {
                health = 40 * level;
                experience = 0;
                dead = true;
                Console.WriteLine("You're dead\n");
            }
        }

        public string GetRace()
        {
            return race;
        }

        public int GetLevel()
        {
            return level;
        }

        public int GetHealth()
        {
            return health;
        }

        public int GetMana()
        {
            return mana;
        }

        public int GetExperience()
        {
            return experience;
        }

        public float GetDamage()
        {
            return damage + GetEquipmentDamage();
        }

        private void UpdateData()
        {
            health = 40 * level;
            mana = 20 * level;
            damage = 5 * level;
        }

        private int GetEquipmentDamage()
        {
            int finalDamage = 0;
            if (leftHand != null)
                finalDamage += leftHand.GetDamage();
            if (rightHand != null)
                finalDamage += rightHand.GetDamage();
            return finalDamage;
        }
    
}
}
